loki/transform/transform_extract_contained_procedures.py

```diff
@@ -0,0 +1,198 @@
+        end subroutine inner
+    is returned.
+    """
+    new_procedures = []
+    for r in procedure.subroutines:
+        new_procedures += [extract_contained_procedure(procedure, r.name)]
+
+    # Remove all subroutines (or functions) from the CONTAINS section.
+    newbody = tuple(r for r in procedure.contains.body if not isinstance(r, Subroutine))
+    procedure.contains = procedure.contains.clone(body = newbody)
+    return new_procedures
+
+def extract_contained_procedure(procedure, name):
+    """
+    Extract a single contained procedure with name ``name`` from the parent procedure ``procedure``.
+
+    This function does the following transforms:
+    1. all global bindings from the point of view of the contained procedure are introduced
+    as imports or dummy arguments to the modified contained procedure returned from this function.
+    2. all calls or invocations of the contained procedure in the parent are modified accordingly.
+
+    See also the "driver" function ``extract_contained_procedures``, which applies this function to each
+    contained procedure of a parent procedure and additionally empties the CONTAINS section of subroutines.
+    """
+    inner = procedure.subroutine_map[name] # Fetch the subprocedure to extract (or crash with 'KeyError').
+
+    # Check if there are variables that don't have a scope. This means that they are not defined anywhere
+    # and execution cannot continue.
+    undefined = tuple(v for v in FindVariables().visit(inner.body) if not v.scope)
+    if undefined:
+        msg = f"The following variables appearing in the contained procedure '{inner.name}' are undefined "
+        msg += f"in both '{inner.name}' and the parent procedure '{procedure.name}': "
+        for u in undefined:
+            msg += f"{u.name}, "
+        raise RuntimeError(msg)
+
+    ## PRODUCING VARIABLES TO INTRODUCE AS DUMMY ARGUMENTS TO `inner`.
+    # Produce a list of variables defined in the scope of `procedure` that need to be resolved in `inner`'s scope
+    # by introducing them as dummy arguments.
+    # The second line drops any derived type fields, don't want them, since want to resolve the derived type itself.
+    vars_to_resolve = [v for v in FindVariables().visit(inner.body) if v.scope == procedure]
+    vars_to_resolve = [v for v in vars_to_resolve if not v.parent]
+
+    # Save any `DeferredTypeSymbol`s for later, they are in fact defined through imports in `procedure`,
+    # and therefore not to be added as arguments to `inner`. (the next step removes them from `vars_to_resolve`)
+    var_imports_to_add = tuple(v for v in vars_to_resolve if isinstance(v, DeferredTypeSymbol))
+
+    # Lookup the definition of the variables in `vars_to_resolve` from the scope of `procedure`.
+    # This provides maximal information on them.
+    vars_to_resolve = [procedure.variable_map[v.name] for v in vars_to_resolve \
+        if v.name in procedure.variable_map.keys()]
```

Comment:
Just to point out that `keys()` is redundant here (and sometimes flagged by linters) and you can mangle the contains check and lookup into one via the walrus operator:

```suggestion
    vars_to_resolve = [proc_var for v in vars_to_resolve if (proc_var := procedure.variable_map.get(v.name))]
```

Reply:
True.